### ISSUE TITLE
Variadic XOR type

### ISSUE THREAD

**Issue by sobolevn** [ts-essentials/ts-essentials#183]
Hi! Thanks a lot for the `XOR` type! It is very useful.

However, currently I have to write a lot of boilerplate to chain 40+ objects with `XOR`.

This is how it looks like:

```ts
export type Task =
    XOR<{ a: AModule },
      XOR<{b: BModule }, 
        XOR<{ c: CModule }, 
          XOR<{ d: DModule }, 
            XOR<{ e: EModule }, { f: FModule }>
          >
        >,
      >
    >;
```

I would love to have somthing like:

```ts
XOR<
  { a: AModule },
  { b: BModule },
  { c: CModule },
  { d: DModule },
  { e: EModule },
  { f: FModule },
  ...
>
```

**Comment by bbarry** [comment]
It is sorta possible now with an abuse of tuple generics but the syntax will involve an extra `[]`:

```typescript
import { XOR } from 'ts-essentials';

type CurryXOR<T, R extends unknown[]> = {
    0: [T],
    1: R extends readonly [infer U, ...infer V]
    ? [...CurryXOR<XOR<T, U>, V>]
    : never
}[R extends readonly [infer _, ...infer __]  ? 1 : 0] ;

type XOR2<R extends unknown[]> = {
    0: R extends readonly [infer U] ? U : never,
    1: R extends readonly [infer U, ...infer V]
    ? CurryXOR<U, V>[0]
    : never
}[R extends readonly [infer _, ...infer __]  ? 1 : 0];

type myType = XOR2<[
  { a: string }, 
  { b: string },
  { c: string },
  { d: string },
  { e: string },
  { f: string }
]>;

const s: myType = {a: ''};

// same as:
type myType2 = XOR<
  XOR<
    XOR<
      XOR<
        XOR<{ a: string }, { b: string }>,
        { c: string }>,
      { d: string }>,
    { e: string }>,
  { f: string }>;


const s2: myType2 = {a: ''};
```

You might be able to simplify that some.

**Comment by sobolevn** [comment]
@bbarry thanks a lot! 

But, this sadly has `Type instantiation is excessively deep and possibly infinite.` problem: [playground](https://www.typescriptlang.org/play?#code/FAegVGAECyCGBOBrSBnA9gWwKaQA7zVy3gBcBLLFSAdwAsyBjWyNAOwBsBPSE2nAMzTxs8SGVYN2AVwAm4gOaRYVQuTax2S1jNRYSPPpABuGqThJpIrLEeKQwIYCU5FIAdTK80UkgB4AKgA0kACqAHyQALyQAN6QANoACmKskACiAB6Ssli+iFicaPyQQZD5hcXhALoA-ABcVjZ2AL4A3MCgEJDyegY4ABoA8gBKPC44dIzMDN7sOhiw+ZAATGNEVFhZ0jI4WLBMLLx2DsCbuEL6zq5DwwHB4VElkAA+oZCbJFjaKgBGAFZYBj6GqQAAUHi8PjuoQiADJQgBKF5giG0bx+ELBfxwkpIhr+ZEhdpOcaQADCUng8E4N2how+XxkVCkrEQrDQ1FY8SqEWiMWAkEFkAADA14v4qoEBUKAIwNekZT7fSDwPYyNhcBLifh2TGQAB0hu1dgAalVpYKQfFDfqKVSaSNfLTSuFgiawuahZAGtZbPBgM14gqlUyVWqNdx4sbRAB9YI26OQGMxqqWyAy70i1PEq4DEbLXzBxnM1nsznc3mxC0i+XvRXFsOwdUcSOJkKpkEhTO+4hSr1yyBF5WqpsRrWsHWiPUJiemz1CkF26m0vXu+LC+eCn1Nf2BoehkfNzVR2ex+NG09JlNpjMNDc50kYTj+UnRG4F+LSuKwBooEjwBRIGaYIv0gH5f3-QDgNAhgIIA1hFGgwU4hkOCoL7WJ3jQhCgIwuJ+GwxC8O6QjcNA2ABz-eCiMgUCfkoyCcKQzCGAY6iyOQyAZDY9DQKwHimOI-gBKI0D5BEsieWJGZWD-VAGifF9XD5H9IAAcjUtogA)

```
```

**Comment by emilien-jegou** [comment]
Hello :wave: 

Reopening this topic with a solution for anyone blocked, essentially you have two ways to do this, either you use [maninak](https://github.com/maninak/ts-xor/blob/master/src/types/xor.ts) implementation which "brute-force" a XOR with 100 template parameters or you use the array version below:

```typescript
// This is not needed but it result in prettier types from LSP
export type Prettify<T> = { [K in keyof T]: T[K] } & {};

type XORArray<Types extends unknown[]> = Prettify<
   Types extends [infer First, ...infer Rest] ?
      Rest extends [infer Second, ...infer Rest] ?
         XOR<XOR<First, Second>, XORArray<Rest>>
         : First
      : {}>;


type M = XORArray<[{ a: true, c: number }, { b: true }, { c: string }]>;

// ... Transpile to:
// type M =
//  | { a?: undefined; b?: undefined; c: string; }
//  | { a?: undefined; c?: undefined; b: true; }
//  | { b?: undefined; a: true; c: number; }
```

The XOR implementation from maninak should be cross compatible with the one in ts-essentials, may be worth testing it and bringing it over.

**Comment by Beraliv** [comment]
> Hello :wave: 
> 
> 
> 
> Reopening this topic with a solution for anyone blocked, the best way to fix this issue is to use [maninak](https://github.com/maninak/ts-xor/blob/master/src/types/xor.ts) implementation which "brute-force" a XOR with 100 template parameters.

I wonder what's the absolute maximum that anyone is currently using? Is it 50, 100 type parameters?

**Comment by Beraliv** [comment]
@emilien-jegou thank you for suggesting the brute-force solution. After thinking through it, I've decided to add 50 generic types

cc @sobolevn it's fixing the problem that you've reported a while ago, please let me know if you find it useful - https://tsplay.dev/NVXa5W

**Comment by Beraliv** [comment]
The feature with up to 50 generic type parameters has been just merged, I will include it in the next release - https://github.com/ts-essentials/ts-essentials/pull/386

**Comment by Beraliv** [comment]
It's released just now as a part of [v10.0.0](https://github.com/ts-essentials/ts-essentials/releases/tag/v10.0.0)

**Comment by papermana** [comment]
This broke type checking for us, see https://github.com/ts-essentials/ts-essentials/issues/400.
